tensorflow/core/tpu/tpu_initializer_helper.cc

```diff
@@ -51,6 +55,55 @@ bool GetEnvBool(const char* name, bool defval) {
 
 }  // namespace
 
+// This function gets pid of a process and checks if that process is using tpu.
+// It is not able to check processes that are owned by another user.
+bool IsTpuUsed(long pid) {
+  std::string path = absl::StrCat("/proc/", pid, "/fd");
+  DIR* raw_fd_dir = opendir(path.c_str());
+  if (!raw_fd_dir) {
+    return false;
+  }
+  std::unique_ptr<DIR, int (*)(DIR*)> fd_dir(raw_fd_dir, closedir);
+  struct dirent* ent;
+  char line[100];
+  while (ent = readdir(raw_fd_dir)) {
+    if (!isdigit(*ent->d_name)) continue;
+    long fd = strtol(ent->d_name, NULL, 10);
+    path = absl::StrCat("/proc/", pid, "/fd/", fd);
+    if (!readlink(path.c_str(), line, 100)) return false;
+    if (strncmp(line, "/dev/accel0", 11) != 0) continue;
```

Comment:
You can avoid using strncmp by making path a std::string. The key trick here is using `std::string::data` (https://en.cppreference.com/w/cpp/string/basic_string/data) to access the underlying char*, so you can write the string using readlink. (I think data() is like c_str() except it doesn't guarantee a null-terminated string, which is fine in this case.) Something like this:
```c++
std::string tpu_dev_path = "/dev/accel0";
std::string line;
line.resize(tpu_dev_path.size());
while (...) {
  ...
  if (!readlink(path.c_str(), line.data(), line.size()) return false;
  if (line != "/dev/accel0") continue;
  return true;
}
```
Note I also made tpu_dev_path a variable so we can make `line` exactly that length, instead of hardcoding 11.

Reply:
I get the conversion from const char* to char* error by just using a string for line. But I did change the line to string after readlink, and then removed the strcmp because it looked nicer!

Reply:
Ah, I incorrectly though data() had a non-const overload. Maybe just use `&line[0]` then: https://stackoverflow.com/questions/1919626/can-i-get-a-non-const-c-string-back-from-a-c-string